From 365528d2fee0d87373a61bdc1558c421448c7ae3 Mon Sep 17 00:00:00 2001
From: Jens Trillmann <jens.trillmann@governikus.de>
Date: Thu, 18 Sep 2025 09:42:52 +0200
Subject: A11y: Prepend new items in ItemView based on modelIndex

When traversing list elements with a screen reader backwards the elements become
unordered as new elements are only appended to the list. By prepending new items when necessary the elements are correctly ordered.

Task-number: QTBUG-140463
Pick-to: 6.10
Change-Id: I76e6bd8247a4d37d4508fb0f79cd39e7594858ca
---
 src/quick/items/qquickitem.cpp     | 13 ++++++++-----
 src/quick/items/qquickitem.h       |  2 +-
 src/quick/items/qquickitem_p.h     |  2 +-
 src/quick/items/qquickitemview.cpp | 10 ++++++++--
 4 files changed, 18 insertions(+), 9 deletions(-)

diff --git x/qtdeclarative/src/quick/items/qquickitem.cpp y/qtdeclarative/src/quick/items/qquickitem.cpp
index 6b511990c3a39f191debc0f4f8ea5cecacad25d3..c2ab792139e4ddf1caaacc78dd071a4ec4938f7a 100644
--- x/qtdeclarative/src/quick/items/qquickitem.cpp
+++ y/qtdeclarative/src/quick/items/qquickitem.cpp
@@ -2748,7 +2748,7 @@ QQuickItem *QQuickItem::parentItem() const
     return d->parentItem;
 }
 
-void QQuickItem::setParentItem(QQuickItem *parentItem)
+void QQuickItem::setParentItem(QQuickItem *parentItem, bool prepend)
 {
     Q_D(QQuickItem);
     if (parentItem == d->parentItem)
@@ -2820,7 +2820,7 @@ void QQuickItem::setParentItem(QQuickItem *parentItem)
         auto oldParentItem = d->parentItem;
         d->parentItem = parentItem;
         if (d->parentItem) {
-            QQuickItemPrivate::get(d->parentItem)->addChild(this);
+            QQuickItemPrivate::get(d->parentItem)->addChild(this, prepend);
             alreadyAddedChild = true;
         }
         if (d->window) {
@@ -2840,7 +2840,7 @@ void QQuickItem::setParentItem(QQuickItem *parentItem)
     d->dirty(QQuickItemPrivate::ParentChanged);
 
     if (d->parentItem && !alreadyAddedChild)
-        QQuickItemPrivate::get(d->parentItem)->addChild(this);
+        QQuickItemPrivate::get(d->parentItem)->addChild(this, prepend);
     else if (d->window && !alreadyAddedChild)
         QQuickWindowPrivate::get(d->window)->parentlessItems.insert(this);
 
@@ -3037,13 +3037,16 @@ QList<QQuickItem *> QQuickItemPrivate::paintOrderChildItems() const
     return childItems;
 }
 
-void QQuickItemPrivate::addChild(QQuickItem *child)
+void QQuickItemPrivate::addChild(QQuickItem *child, bool prepend)
 {
     Q_Q(QQuickItem);
 
     Q_ASSERT(!childItems.contains(child));
 
-    childItems.append(child);
+    if (prepend)
+        childItems.prepend(child);
+    else
+        childItems.append(child);
 
     QQuickItemPrivate *childPrivate = QQuickItemPrivate::get(child);
 
diff --git x/qtdeclarative/src/quick/items/qquickitem.h y/qtdeclarative/src/quick/items/qquickitem.h
index 7ce93846206b3bf5da68bd00845dea577830172a..73b1eb1ebf3c47e506aad0c9171a0e5c6a41268d 100644
--- x/qtdeclarative/src/quick/items/qquickitem.h
+++ y/qtdeclarative/src/quick/items/qquickitem.h
@@ -183,7 +183,7 @@ public:
 
     QQuickWindow *window() const;
     QQuickItem *parentItem() const;
-    void setParentItem(QQuickItem *parent);
+    void setParentItem(QQuickItem *parent, bool prepend = false);
     void stackBefore(const QQuickItem *);
     void stackAfter(const QQuickItem *);
 
diff --git x/qtdeclarative/src/quick/items/qquickitem_p.h y/qtdeclarative/src/quick/items/qquickitem_p.h
index aba872c5ffc11eefb99b048b39c74fac1aae9d82..6f322d72ade819f8db14c4915c977e29f06ca028 100644
--- x/qtdeclarative/src/quick/items/qquickitem_p.h
+++ y/qtdeclarative/src/quick/items/qquickitem_p.h
@@ -622,7 +622,7 @@ public:
     QList<QQuickItem *> childItems;
     mutable QList<QQuickItem *> *sortedChildItems;
     QList<QQuickItem *> paintOrderChildItems() const;
-    void addChild(QQuickItem *);
+    void addChild(QQuickItem *, bool prepend = false);
     void removeChild(QQuickItem *);
     void siblingOrderChanged();
 
diff --git x/qtdeclarative/src/quick/items/qquickitemview.cpp y/qtdeclarative/src/quick/items/qquickitemview.cpp
index c2adf10eff872cadc7705758b058e3fb47712a19..3efbdeafc406564cfe646e066380544830afe46c 100644
--- x/qtdeclarative/src/quick/items/qquickitemview.cpp
+++ y/qtdeclarative/src/quick/items/qquickitemview.cpp
@@ -2583,13 +2583,19 @@ void QQuickItemView::createdItem(int index, QObject* object)
     }
 }
 
-void QQuickItemView::initItem(int, QObject *object)
+void QQuickItemView::initItem(int index, QObject *object)
 {
+    Q_D(QQuickItemView);
     QQuickItem* item = qmlobject_cast<QQuickItem*>(object);
     if (item) {
         if (qFuzzyIsNull(item->z()))
             item->setZ(1);
-        item->setParentItem(contentItem());
+        FxViewItem* firstItem = d->firstItemInView();
+        bool prepend = false;
+        if (firstItem)
+            prepend = index < firstItem->index;
+
+        item->setParentItem(contentItem(), prepend);
         QQuickItemPrivate::get(item)->setCulled(true);
     }
 }
